service/api_patch.go

```diff
@@ -215,74 +60,41 @@ func (as *APIServer) submitPatch(w http.ResponseWriter, r *http.Request) {
 			as.LoggedError(w, r, http.StatusBadRequest, err)
 			return
 		}
-		if len(data.Patch) > patch.SizeLimit {
-			as.LoggedError(w, r, http.StatusBadRequest, errors.New("Patch is too large."))
-		}
-		finalize = data.Finalize
-
-		apiRequest = PatchAPIRequest{
-			ProjectId:     data.Project,
-			ModuleName:    r.FormValue("module"),
-			Githash:       data.Githash,
-			PatchContent:  data.Patch,
-			BuildVariants: strings.Split(data.Variants, ","),
-			Tasks:         data.Tasks,
-			Description:   data.Description,
+		finalize := data.Finalize
+		variants := strings.Split(data.Variants, ",")
+
+		var err error
+		intent, err = patch.NewCliIntent(dbUser.Id, data.Project, data.Githash, r.FormValue("module"), data.Patch, data.Description, finalize, variants, data.Tasks)
+		if err != nil {
+			as.LoggedError(w, r, http.StatusBadRequest, err)
+			return
 		}
 	}
 
-	project, patchDoc, err := apiRequest.CreatePatch(
-		finalize, as.Settings.Credentials["github"], dbUser, &as.Settings)
-	if err != nil {
-		as.LoggedError(w, r, http.StatusBadRequest, errors.Wrap(err, "Invalid patch"))
+	if err := intent.Insert(); err != nil {
+		as.LoggedError(w, r, http.StatusBadRequest, err)
 		return
 	}
 
-	//expand tasks and build variants and include dependencies
-	if len(patchDoc.BuildVariants) == 1 && patchDoc.BuildVariants[0] == "all" {
-		patchDoc.BuildVariants = []string{}
-		for _, buildVariant := range project.BuildVariants {
-			if buildVariant.Disabled {
-				continue
-			}
-			patchDoc.BuildVariants = append(patchDoc.BuildVariants, buildVariant.Name)
-		}
-	}
-
-	if len(patchDoc.Tasks) == 1 && patchDoc.Tasks[0] == "all" {
-		patchDoc.Tasks = []string{}
-		for _, t := range project.Tasks {
-			if t.Patchable != nil && !(*t.Patchable) {
-				continue
-			}
-			patchDoc.Tasks = append(patchDoc.Tasks, t.Name)
-		}
+	patchID := bson.NewObjectId()
+	job := units.NewPatchIntentProcessor(patchID, intent)
+	if err := as.queue.Put(job); err != nil {
+		as.LoggedError(w, r, http.StatusInternalServerError, err)
+		return
 	}
-
-	var pairs []model.TVPair
-	for _, v := range patchDoc.BuildVariants {
-		for _, t := range patchDoc.Tasks {
-			if project.FindTaskForVariant(t, v) != nil {
-				pairs = append(pairs, model.TVPair{v, t})
-			}
-		}
+	if !amboy.WaitJobInterval(job, as.queue, 10*time.Second) {
```

Comment:
this is going to block for the queue (forever). there's some internal polling that happens every 10 seconds (so this will take some multiple of 10 seconds, not have a 10 second deadline.) 

You probably want amboy.WaitCtxJobInterval() or something similar to have a timeout. 

Having said that, maybe it makes sense to just call run on the job method and check for errors locally: if you're not actually going to let the job run asynchronously, then we might as well skip the queue all together.

Reply:
done, running the job synchronously.